### PR TITLE
Catching errors on grpc and adding closing grpc channel

### PR DESCRIPTION
**Pull request by noiarek**
**What changed?**
Three things

1. Adding clearer variables to mappers (sorry for the noise on review time)
2. follow up catching rpc error exception and returning a general Exception for error so far
3. Closing grpc channel and adding method to tchannel too
4. follow up better assertion on list domains test

**Why?**
1. Cleaner code
2. So far it was just throwing a general exception and not following the `result, error` pattern
3. channels were remained open

**How did you test it?**
1. Running the client

**Potential risks**
Nothing (maybe)

 
**Release notes**
No

### CODE REVIEW

File: cadence/gateway/cadence/grpc_service.py [hunk 1]
```diff
@@ -22,25 +29,36 @@ def __init__(self, host: str = 'localhost', port: int = 7833, timeout: int = 300
...
             '{}:{}'.format(host, port))
-
         # stubs
         self.domain = service_domain_pb2_grpc.DomainAPIStub(self.channel)
         self.service_workflow = service_workflow_pb2_grpc.WorkflowAPIStub(self.channel)
 
+    def close(self):
+        self.channel.close()
+
+    def set_next_timeout_cb(self, cb: Callable):
```
Comment (longquanzheng) [inline cadence/gateway/cadence/grpc_service.py:39]: Do you know what is this for?

File: cadence/errors.py [hunk 2]
```diff
@@ -106,6 +106,12 @@ class ClientVersionNotSupportedError(Exception):
     supported_versions: str
 
 
+@dataclass
+class RPCError(Exception):
+    message: str
+    code: grpc.StatusCode
```
Comment (longquanzheng) [inline cadence/errors.py:112]: I think the idea of this file is to define types independent of thrift/grpc. So it's better to not import any grpc types.

File: cadence/gateway/cadence/grpc_service.py [hunk 3]
```diff
@@ -22,25 +29,36 @@ def __init__(self, host: str = 'localhost', port: int = 7833, timeout: int = 300
...
\ No newline at end of file
+        try:
+            response = self.service_workflow.StartWorkflowExecution.with_call(
+                grpc_request,
+                metadata=self.metadata,
+                timeout=self.timeout
+            )
+            return start_workflow_execution_response_to_dataclass(response[0]), None
+        except grpc.RpcError as e:
+            return None, process_error(e)
```
Comment (longquanzheng) [inline cadence/gateway/cadence/grpc_service.py:64]: how does it convert to error `WorkflowExecutionAlreadyStartedError` ?

Comment (longquanzheng) [reply cadence/gateway/cadence/grpc_service.py:64]: If starting same workflowID twice, you should see the error.

File: cadence/errors.py (hunk 2 above)
Comment (noiarek) [reply cadence/errors.py:112]: Makes sense. I'll start figuring out how to do the mapping

File: cadence/gateway/cadence/grpc_service.py (hunk 1 above)
Comment (noiarek) [reply cadence/gateway/cadence/grpc_service.py:39]: no, this is a copy paste from the tchannel version.

File: cadence/gateway/cadence/grpc_service.py (hunk 3 above)
Comment (noiarek) [reply cadence/gateway/cadence/grpc_service.py:64]: Screen Shot 2021-12-23 at 12 12 32 PM

Comment (noiarek) [reply cadence/gateway/cadence/grpc_service.py:64]: Do you know where I can find all the status code so I can do the mapping for error?

Comment (noiarek) [reply cadence/gateway/cadence/grpc_service.py:64]: Forget my latest comment. I think I know how to do this mapping. Thanks!

File: cadence/grpc_errors.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+from cadence.errors import WorkflowExecutionAlreadyStartedError, ServiceBusyError, QueryFailedError, LimitExceededError, \
+    CancellationAlreadyRequestedError, ClientVersionNotSupportedError, DomainAlreadyExistsError, DomainNotActiveError, \
+    EntityNotExistsError, WorkflowExecutionAlreadyCompletedError, FeatureNotEnabledError
+from uber.cadence.api.v1 import error_pb2
+
+
+def process_error(e: grpc.RpcError) -> Exception:
+    status = rpc_status.from_call(e)
+    for detail in status.details:
```
Comment (noiarek) [inline cadence/grpc_errors.py:12]: There were some errors like `BadRequestError` or `AccessDeniedError` that were not present in the proto file, however I found them in the go client https://github.com/uber-go/cadence-client/blob/master/internal/compatibility/proto/error.go#L36

I'll figure out how to trigger them, how it looks as proto error and I'll map it later. I suspect they might land in a more general category defined by grpc already

File: cadence/gateway/cadence/grpc_service.py (hunk 3 above)
Comment (noiarek) [reply cadence/gateway/cadence/grpc_service.py:58]: done 👍

**Comment by noiarek** [comment]
Closing since the other diff was approved

**Pull request closed without merging**